operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+        - TODO: Use a worker pool to process the queue as quickly as possible, rather than waiting for poll cycles to trigger batching.
+3. Opening
+    1. Each of the matched files is opened. Note:
+        - A small amount of time has passed since the file was matched.
+        - It is possible that is has been moved or deleted by this point.
+        - Only a minimum set of operations should occur between file matching and opening.
+        - If an error occurs while opening, it is logged.
+4. Fingerprinting
+    1. The first `N` bytes of each file are read. (See fingerprinting section above.)
+5. Exclusion
+    1. Empty files are closed immediately and discarded. (There is nothing to read.)
+    2. Fingerprints found in this batch are cross referenced against each other to detect duplicates. Duplicate files are closed immediately and discarded.
+        - In the vast majority of cases, this occurs during file rotation that uses the copy/truncate method. (See fingerprinting section above.)
+6. Reader Creation
+    1. Each file handle is wrapped into a `Reader` along with some metadata. (See Reader section above)
+        - During the creation of a `Reader`, the file's fingerprint is crossreferenced with previously known fingerprints.
+        - If a file's fingerprint matches one that has recently been seen, then metadata is copied over from the previous iteration of the Reader. Most importantly, the offset is accurately maintained in this way.
+        - If a file's fingerprint does not match any recently seen files, then its offset is initialized according to the `start_at` setting.
+7. Detection of "Lost" Files
+8. Consumption
+    1. Lost files are consumed. Since these files are thought to have been rotated to a location where we will not match them again, we will never see them again. 
+        - The best we can do is finish consuming their current contents.
+        - We can reasonable expect in most cases that these files are no longer being written to.
+    2. Matched files (from this poll cycle) are consumed.
+        - These file handles will be left open until the end of the next poll cycle.
+        - Typically, we can expect to find most of these files again. However, these files are consumed greedily, in case we do not see them again.
+    3. All files in both sets are consumed concurrently.
+9. Closing
+    1. All files from the previous poll cycle are closed.
+10. Archiving
+    1. Readers created in the current poll cycle are added to the historical record.
+11. Pruning
+    1. The historical record is purged of Readers that have existsed for 3 generations.
+        - This number is somewhat arbitrary, and should probably be made configurable. However, its exact purpose is quite obscure.
+12. Persistence
+    1. The historical record of readers is synced to whatever persistence mechanism was provided to the operator.
+13. End Poll Cycle
+    1. At this point, the operator sits idle until the poll timer fires again.
```

Comment:
Would it make more sense to eliminate the idling if the queue is not empty so that we do the remaining work asap?

Reply:
Yeah, I actually realized the same when drafting this and had added a TODO in the `Queueing` section. Basically, I think we could use a worker pool to consume the queue.